evennia/contrib/cooldowns.py

```diff
@@ -0,0 +1,188 @@
+
+    @property
+    def all(self):
+        """
+        Returns a list of all keys in this object.
+        """
+        return list(self.data.keys())
+
+    def ready(self, *args):
+        """
+        Checks whether all of the provided cooldowns are ready (expired).
+        If a requested cooldown does not exist, it is considered ready.
+
+        Args:
+            any (str): One or more cooldown names to check.
+        Returns:
+            (bool): True if each cooldown has expired or does not exist.
+        """
+        return self.time_left(*args) <= 0
+
+    def time_left(self, *args):
+        """
+        Returns the maximum amount of time left on one or more given
+        cooldowns. If a requested cooldown does not exist, it is
+        considered to have 0 time left.
+
+        Args:
+            any (str): One or more cooldown names to check.
+        Returns:
+            (int): Number of seconds until all provided cooldowns are
+                ready. Returns 0 if all cooldowns are ready (or don't
+                exist.)
+        """
+        now = time.time()
+        cooldowns = [self.data[x] - now for x in args if x in self.data]
+        if not cooldowns:
+            return 0
+        return math.ceil(max(max(cooldowns), 0))
+
+    def set(self, cooldown, seconds):
+        """
+        Sets a given cooldown to last for a specific amount of time.
+
+        If this cooldown is already set, this replaces it.
+
+        Args:
+            cooldown (str): The name of the cooldown.
+            seconds (int): The number of seconds before this cooldown is
+                ready again.
+        """
+        now = time.time()
+        self.data[cooldown] = int(now) + (max(seconds, 0) if seconds else 0)
+
+    def extend(self, cooldown, seconds):
+        """
+        Adds a specific amount of time to an existing cooldown.
+
+        If this cooldown is already ready, this is equivalent to calling
+        set. If the cooldown is not ready, it will be extended by the
+        provided duration.
+
+        Args:
+            cooldown (str): The name of the cooldown.
+            seconds (int): The number of seconds to extend this cooldown.
+        """
+        time_left = self.time_left(cooldown)
+        self.set(cooldown, time_left + (seconds if seconds else 0))
+
+    def reset(self, cooldown):
+        """
+        Resets a given cooldown.
+
+        Args:
+            cooldown (str): The name of the cooldown.
+        """
+        if cooldown in self.data:
+            del self.data[cooldown]
+
+    def clear(self):
+        """
+        Resets all cooldowns.
+        """
+        for cooldown in list(self.data.keys()):
+            del self.data[cooldown]
+
+    def cleanup(self):
+        """
+        Deletes all expired cooldowns. This helps keep attribute storage
+        requirements small.
+        """
+        now = time.time()
+        keys = [x for x in self.data.keys() if self.data[x] - now < 0]
```

Comment:
Rather than looping over the (Attribute) data set, which means a lot of deserializations and re-saving to the database, I'd suggest extracting the data and decouple it from the database at the top of this method. Then loop+delete over the in-memory dict. Afterwards you save the resulting dict back to the Attribute.

This matters little for most code, but when you are making a formal thing like this, one can just as well do it efficiently. :wink:

Reply:
Done. Note that I had to add storage for the parent object and db attribute name to enable this. I changed this class to use __slots__ to minimize memory impact.